patches/server/0928-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,1156 @@
++import io.papermc.paper.command.brigadier.CommandSourceStack;
++import net.minecraft.commands.synchronization.ArgumentTypeInfos;
++
++import java.util.function.Predicate;
++
++/*
++Basically this converts the argument to a different type when parsing.
++ */
++public class WrappedArgumentCommandNode<NMS, API> extends ArgumentCommandNode<CommandSourceStack, NMS> {
++
++    private final ArgumentType<API> argument;
++
++    public WrappedArgumentCommandNode(String name, ArgumentType<API> argument, ArgumentType<NMS> nms, Command<CommandSourceStack> command, Predicate<CommandSourceStack> requirement, CommandNode<CommandSourceStack> redirect, RedirectModifier<CommandSourceStack> modifier, boolean forks, SuggestionProvider<CommandSourceStack> customSuggestions) {
++        super(name, nms, command, requirement, redirect, modifier, forks, customSuggestions);
++        if (!ArgumentTypeInfos.isClassRecognized(nms.getClass())) {
++            // Is this argument an NMS argument?
++            throw new IllegalArgumentException("Unexpected argument type was passed. This should be an NMS type!");
++        }
++
++        this.argument = argument;
++    }
++
++    // See ArgumentCommandNode#parse
++    @Override
++    public void parse(StringReader reader, CommandContextBuilder<CommandSourceStack> contextBuilder) throws CommandSyntaxException {
++        final int start = reader.getCursor();
++        API result = this.argument.parse(reader); // Use the api argument parser
++        final ParsedArgument<CommandSourceStack, API> parsed = new ParsedArgument<>(start, reader.getCursor(), result); // Return an API parsed argument instead.
++
++        contextBuilder.withArgument(this.getName(), parsed);
++        contextBuilder.withNode(this, parsed.getRange());
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigForwardingMap.java b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigForwardingMap.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..c8889e5ae6d7283eb80f40cd59b7e3d682ade0f5
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitBrigForwardingMap.java
+@@ -0,0 +1,194 @@
++package io.papermc.paper.command.brigadier.bukkit;
++
++import com.mojang.brigadier.CommandDispatcher;
++import com.mojang.brigadier.tree.CommandNode;
++import com.mojang.brigadier.tree.LiteralCommandNode;
++import io.papermc.paper.command.brigadier.CommandSourceStack;
++import io.papermc.paper.command.brigadier.bukkit.BukkitCommandNode;
++import org.bukkit.Bukkit;
++import org.bukkit.command.Command;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.ArrayList;
++import java.util.Collection;
++import java.util.HashSet;
++import java.util.List;
++import java.util.Map;
++import java.util.Set;
++
++public class BukkitBrigForwardingMap implements Map<String, Command> {
```

Comment:
Maybe this should extend HashMap? or whatever concrete type it was before? Idk if somethings casted to HashMap

Reply:
Oh truth.... I guess I'll just override the getters/put method logic.